text/0031-corejam.md

```diff
@@ -0,0 +1,681 @@
+
+An *Authorizer* is a parameterized procedure:
+
+```rust
+type CodeHash = [u8; 32];
+type AuthParamSize = ConstU32<1024>;
+type AuthParam = BoundedVec<u8, AuthParamSize>;
+struct Authorizer {
+    code_hash: CodeHash,
+    param: AuthParam,
+}
+```
+
+The `code_hash` of the Authorizer is assumed to be the hash of some code accessible in the Relay-chain's Storage pallet. The procedure itself is called the *Authorization Procedure* (`AuthProcedure`) and is expressed in this code (which must be capable of in-core VM execution). Its entry-point prototype is:
+
+```rust
+fn is_authorized(param: &AuthParam, package: &WorkPackage, core_index: CoreIndex) -> bool;
+```
+
+This function is executed in a metered VM and subject to a modest system-wide limitation on execution time. If it overruns this limit or panicks on some input, it is considered equivalent to returning `false`. While it is mostly stateless (e.g. isolated from any Relay-chain state) it is provided with the package's `context` field in order to give information about a recent Relay-chain block. This allows it to be provided with a concise proof over some recent Relay-chain state.
+
+A single `Authorizer` value is associated with the index of the Core at a particular Relay-chain block and limits in some way what Work Packages may be legally processed by that Core.
+
+Since encoded `Authorizer` values may be fairly large (up to 1,038 bytes here), they may not be a drop-in replacement for the `ParaId`/`TaskId` used at present in the Agile Coretime interface. Because of this, we provide a lookup mechanism allowing a much shorter `AuthId` to be used within the Coretime scheduling messaging. Conveniently, this is precisely the same datatype size (32-bit) as a `ParaId`/`TaskId`.
+
+There is an Authorizations Pallet which stores the association. Adding a new piece of code is permissionless but requires a deposit commensurate with its size.
+
+```rust
+type AuthId = u32;
+type Authorizers = StorageMap<AuthId, Authorizer>;
+```
+
+An *Authorization* is simply a blob which helps the Authorizer recognize a properly authorized Work Package. No constraints are placed on Authorizers over how they may interpret this blob. Expected authorization content includes signatures, Merkle-proofs and more exotic succinct zero-knowledge proofs.
+
+_(Note: depending on future Relay-chain Coretime scheduling implementation concerns, a window of Relay-chain blocks)._
```

Comment:
To where do this note refer?

It's anyways important that work packages cannot know in advance when they'll be backed, when the resulting candidate reciepts get reported on-chain, when they'll be included/integration, or what chain forks containing either those events gets approved and finalized.  

We could set future cut offs for reporting, but not too optimistically near term.   Async backing matters because latency arises from various sources.